<a href="https://colab.research.google.com/github/xDanilog27/Bussiness-Inteligent/blob/main/Reto_4_Modelo_de_Clasificaci%C3%B3_Naive_Bayes_Reto_Parcial_I_M%C3%A9todos_de_Clasificaci%C3%B3n_(Machine_Learning).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Cargar las librerias



In [ ]:
import numpy as np #Posee Funciones Matematicas
import pandas as pd #Comunicacion excel
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import accuracy_score # comparar las predicciones del modelo

# Librerias Arbol
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

from sklearn.cluster import KMeans
from sklearn.metrics import confusion_matrix




1. Cargar la base de datos

In [ ]:
nxl='/content/4. BD1_Longitudinal Survey_L_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)

# Seleccionar las variables
XDB=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income','HealthPlan']]

XDB = XDB.dropna()
display(XDB)

XD=XDB[['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income']]
yd=XDB[['HealthPlan']]




Age  Mother_Edu  Father_Edu  FamilySize  Self_Esteem  Height  Weight  \
1       20         5.0         8.0           5         16.0    62.0   120.0   
5       19        12.0        12.0           4         27.0    64.0   200.0   
6       15        12.0        12.0           3         26.0    65.0   131.0   
7       21         9.0         6.0           3         23.0    65.0   179.0   
8       16        12.0        10.0           6         26.0    66.0   145.0   
...    ...         ...         ...         ...          ...     ...     ...   
12557   20        12.0        12.0           5         20.0    60.0   110.0   
12565   20        12.0        11.0           2         24.0    63.0   152.0   
12647   21        16.0        12.0           1         30.0    70.0   180.0   
12651   21        16.0        16.0           1         27.0    73.0   170.0   
12678   21         8.0         8.0           1         19.0    70.0   180.0   

        Income  HealthPlan  
1          0.0         1.0  
5      40000.0         1.0  
6      25000.0         1.0  
7      27400.0         1.0  
8      52000.0         1.0  
...        ...         ...  
12557   1000.0         1.0  
12565  38000.0         1.0  
12647  43600.0         1.0  
12651  40000.0         1.0  
12678  17000.0         1.0  

[5963 rows x 9 columns]

Modelo

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB() #Aloja el Modelo
mnb.fit(XD,yd) #Entrena el Modelo

#Obtener informacion del modelo
u=mnb.theta_ #Media de los datos
print("La media de los datos es:\n",u) #La primera es No Y la Segunda es Si, en base al Highway

varianza=mnb.var_ #Varianza de los datos
print("La varianza de los datos es:\n",varianza)

sigma=np.sqrt(varianza) #Desviacion Estandar - Se obtiene la desviacion estandar de los ingresos
print("La desviacion estandar de los datos es:\n",sigma)

# Limites superior y limite inferior
LS=u+sigma
LI=u-sigma
print("El limite superior es:\n",LS)
print("El limite inferior es:\n",LI)

XDS=[[18,12,12,5,28,71,180,48000]]
ydp=mnb.predict(XDS)
print("La prediccion es:\n",ydp)

if ydp==0:
  print("No posee un plan de salud")
else:
  print("Si posee un plan de salud")


La media de los datos es:
 [[1.82731536e+01 1.00539273e+01 9.79835873e+00 4.98124267e+00
  2.15439625e+01 6.68968347e+01 1.46289566e+02 1.64244736e+04]
 [1.83774951e+01 1.12438356e+01 1.12348337e+01 4.64559687e+00
  2.26039139e+01 6.69878669e+01 1.45515264e+02 3.52898006e+04]]
La varianza de los datos es:
 [[6.10912942e+00 1.23953711e+01 1.72673504e+01 6.84106924e+00
  1.74740136e+01 1.73067452e+01 8.64167889e+02 4.63341579e+08]
 [6.11179744e+00 1.08478775e+01 1.62968048e+01 5.94083148e+00
  1.76028952e+01 1.74977927e+01 9.15983323e+02 1.31875123e+09]]
La desviacion estandar de los datos es:
 [[2.47166531e+00 3.52070605e+00 4.15540015e+00 2.61554378e+00
  4.18019301e+00 4.16013764e+00 2.93967326e+01 2.15253706e+04]
 [2.47220497e+00 3.29361161e+00 4.03693012e+00 2.43738210e+00
  4.19558044e+00 4.18303630e+00 3.02652164e+01 3.63146146e+04]]
El limite superior es:
 [[2.07448189e+01 1.35746334e+01 1.39537589e+01 7.59678645e+00
  2.57241555e+01 7.10569723e+01 1.75686299e+02 3.79498442e+04]


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


In [ ]:
#1. Numero de datos de la base de datos original
total_data = len(XDB)
print(f"Número total de datos de toda la base de datos: {total_data}")


#2.	¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?

highway_counts=XDB['HealthPlan'].value_counts()
percentage_highway=(highway_counts[1] / total_data) * 100
percentage_no_highway=(highway_counts[0] / total_data) * 100
print(f"No posee un plan de salud: {percentage_highway:.2f}%")
print(f"Si posee un plan de salud: {percentage_no_highway:.2f}%")


Número total de datos de toda la base de datos: 5963
No posee un plan de salud: 85.70%
Si posee un plan de salud: 14.30%


Modelo

In [ ]:
from sklearn.metrics import confusion_matrix
#Pronostico
ybp2=mnb.predict(XD)
cm=confusion_matrix(yd,ybp2) #Crea matriz de confusion
print("La matriz de confusion es:\n", cm)
VN=cm[0,0] #Verdaderos negativos
VP=cm[1,1] #Verdaderos positivos
FN=cm[1,0] #Falsos negativos
FP=cm[0,1] #Falsos positivos

#6
Exactitud=(VP+VN)/(VP+VN+FP+FN)
print("La exactitud es:\n",Exactitud)

#7
Especificidad=VN/(VN+FP)
print("La especificidad es:\n",Especificidad)

#8 y 9
print(f"Verdaderos Negativos: {VN}")
print(f"Verdaderos Positivos: {VP}")

#10
TasaError=(FP+FN)/(VP+VN+FP+FN)
print("La tasa de error es:\n",TasaError)

#11
Sensibilidad=VP/(VP+FN)
print("La sensibilidad es:\n",Sensibilidad)

#12
Precision=VP/(VP+FP)
print("La precición es:\n",Precision)

#13 18,12,12,5,28,71,180,48000
print("La prediccion es:\n",ydp)

if ydp==0:
  print("No posee un plan de salud")
else:
  print("Si posee un plan de salud")

La matriz de confusion es:
 [[ 113  740]
 [ 265 4845]]
La exactitud es:
 0.8314606741573034
La especificidad es:
 0.1324736225087925
Verdaderos Negativos: 113
Verdaderos Positivos: 4845
La tasa de error es:
 0.16853932584269662
La sensibilidad es:
 0.9481409001956947
La precición es:
 0.8675022381378693
La prediccion es:
 [1.]
Si posee un plan de salud


# Analisis de resultados

El análisis se enfoca en la predicción de la variable "Health Plan" (poseer o no un plan de salud) utilizando un modelo de clasificación Naive Bayes. Se emplearon las variables de entrada Age, Mother_Edu, Father_Edu, FamilySize, Self_Esteem, Height, Weight e Income.

El modelo Naive Bayes alcanza una exactitud del 83.14%, lo que indica que clasifica correctamente aproximadamente 8 de cada 10 individuos. Sin embargo, un análisis más detallado revela diferencias en el rendimiento según la categoría predicha.

Especificidad: 13.25%. El modelo tiene dificultades para identificar correctamente a las personas que no poseen un plan de salud, clasificando erróneamente a muchos de ellos como si lo tuvieran.

Sensibilidad: 94.81%. Esto indica que el modelo es muy efectivo para identificar a quienes sí poseen un plan de salud.

Precisión: 86.75%. La mayoría de las predicciones de que una persona tiene un plan de salud son correctas.

Tasa de error: 16.85%. Aproximadamente 1 de cada 6 predicciones es incorrecta.


Utilizando el modelo Naive Bayes, se evaluó a una persona con las siguientes características socioeconómicas: [18, 12, 12, 5, 28, 71, 180, 48000]

El modelo predice que esta persona posee un plan de salud (categoría 1).